Add Trigger and Subscription modal in topology

karthikjeeyar opened the pull request:
**Fixes:** https://issues.redhat.com/browse/ODC-4246 , https://issues.redhat.com/browse/ODC-4167

**Problem:**
Ability to add trigger/subscription from topology gestures and kebab actions

**Solution:**
Added Trigger modal and kebab actions for triggers.
Added Subscription modal and kebab actions for subscriptions
Added context menu on the resource backed edges


**Note: Sidebar enhancements will be added in a follow up PR**

**Screenshot:**
Trigger Modal:

![image](https://user-images.githubusercontent.com/9964343/88234703-0ab54000-cc97-11ea-9522-2ebbb7c85caa.png)

Subscription Modal:

![image](https://user-images.githubusercontent.com/9964343/88234787-320c0d00-cc97-11ea-85a6-7842b157d7fb.png)

Empty state:

![image](https://user-images.githubusercontent.com/9964343/88234890-6253ab80-cc97-11ea-8d73-1cf2728b9567.png)

Error state:
![image](https://user-images.githubusercontent.com/9964343/88235676-fa9e6000-cc98-11ea-8cc3-2f9e4da0549d.png)

Gif:
![brokers-triggers-modal](https://user-images.githubusercontent.com/9964343/88633335-743bb100-d0d2-11ea-931d-b41b2ac0debb.gif)



cc: @rachael-phillips @beaumorley @openshift/team-devconsole-ux

karthikjeeyar commented:
/assign @invincibleJai

karthikjeeyar commented:
/retest

serenamarie125 commented:
@rachael-phillips can you confirm?  I think that the label on the Add Trigger modal should be "Subscriber", not "Select Subscriber"

serenamarie125 commented:
Can we have the Attribute/Value <delete icon> be the same width as the other input widgets?  Right now there's extra space between the delete icon & the edge of the modal.

karthikjeeyar commented:
> @rachael-phillips can you confirm? I think that the label on the Add Trigger modal should be "Subscriber", not "Select Subscriber"

@serenamarie125 @rachael-phillips  Like we have in Design doc, I have changed the title to `Subscriber` and updated the screenshots. The filter attributes section in the trigger modal uses the `nameValueEditor` field that we use every where in console has this type of spacing.
eg: ApiServersource, sinkbinding event source forms

![image](https://user-images.githubusercontent.com/9964343/88254315-88e30800-ccd2-11ea-93fe-b613b01fd92c.png)

rachael-phillips commented:
Hey @karthikjeeyar @serenamarie125 I can confirm that the recommendation to have Subscriber rather than Select Subscriber is correct. I think we should eventually update the spacing of the Resource form fields across the flow, but agree with keeping it standard for now rather than updating based on what I've seen in other places. For example, this is another modal on the topology view, and for consistency I think we should go with the standard spacing. 
<img width="727" alt="Screen Shot 2020-07-23 at 9 50 05 AM" src="https://user-images.githubusercontent.com/40275806/88294239-02c2c380-ccca-11ea-8bd4-30ad4477610d.png">

@serenamarie125 I will add your recommendation to this Audit I am currently working on of our form fields so we can address this in the future: https://docs.google.com/document/d/10QMFBrCGmwbq92Vy8GSNteAX6EgslTcR4PnMoQwF_1w/edit?usp=sharing

christianvogt commented on frontend/packages/dev-console/src/components/topology/components/edges/BaseEdge.tsx:
```diff
@@ -17,7 +18,8 @@ type BaseEdgeProps = {
   className?: string;
-} & WithRemoveConnectorProps;
+} & WithRemoveConnectorProps &
+  WithContextMenuProps;
```
If you use `Partial` then the components that use the `BaseEdge` do not necessarily need to provide context menu props.
```suggestion
  Partial<WithContextMenuProps>;
```

christianvogt commented on frontend/packages/dev-console/src/components/topology/components/edges/TrafficConnector.tsx:
```diff
@@ -1,14 +1,14 @@
 
-const TrafficConnector: React.FC<TrafficConnectorProps> = ({ element }) => (
-  <BaseEdge element={element} className="odc-traffic-connector">
+const TrafficConnector: React.FC<TrafficConnectorProps> = ({ element, ...others }) => (
```
Shouldn't need to change this component if `BaseEdge` props are defined properly.

christianvogt commented on frontend/packages/knative-plugin/src/actions/add-subscription.tsx:
```diff
@@ -15,3 +21,9 @@ export const addSubscription = (model: K8sKind, source: K8sResourceKind): KebabO
+
+export const addSubscriptionWithIcon = (source): KebabOption => {
+  const subscriptionMenu = addSubscription(EventingSubscriptionModel, source);
+  subscriptionMenu.icon = <StarIcon />;
```
Remove the star icon. Fairly certain it was agreed now to have no icons.

cc @openshift/team-devconsole-ux

karthikjeeyar replied:
Yes, i have removed it.

christianvogt commented on frontend/packages/knative-plugin/src/actions/add-trigger.tsx:
```diff
@@ -15,3 +21,9 @@ export const addTrigger = (model: K8sKind, source: K8sResourceKind): KebabOption
+
+export const addTriggerWithIcon = (source): KebabOption => {
+  const triggerMenu = addTrigger(EventingTriggerModel, source);
+  triggerMenu.icon = <StarIcon />;
```
Same comment.
No icon.

christianvogt commented on frontend/packages/knative-plugin/src/components/pub-sub/form-fields/PubSubSubscriber.tsx:
```diff
@@ -0,0 +1,65 @@
+      {resourceAlert && (
+        <>
+          <Alert variant="default" title="No Subscriber available" isInline>
+            To create a Subscriber create a Knative Service from the Add page.
```
```suggestion
            To create a Subscriber create, a Knative Service from the Add page.
```

christianvogt commented on frontend/packages/knative-plugin/src/topology/components/edges/EventSourceLink.tsx:
```diff
@@ -16,7 +17,8 @@ type EventSourceLinkProps = {
 } & WithSourceDragProps &
-  WithTargetDragProps;
+  WithTargetDragProps &
+  WithContextMenuProps;
```
Changes here aren't necessary if you type `BaseEdge` correctly.

christianvogt commented on frontend/packages/knative-plugin/src/topology/components/edges/TrafficLink.tsx:
```diff
@@ -1,14 +1,19 @@
 type TrafficLinkProps = {
   element: Edge;
-};
+} & WithContextMenuProps;
```
Changes here aren't necessary if you type BaseEdge correctly.

karthikjeeyar replied:
updated

christianvogt commented:
@karthikjeeyar ran into an issue where if I set only the value of a filter and then submit the form, I get an error and then the filters key, value, and Add button go missing. 

![image](https://user-images.githubusercontent.com/14068621/88431838-1ef95a00-cdc9-11ea-971a-14e8ca79f533.png)

christianvogt commented:
The rest seems to function well.

invincibleJai commented on frontend/packages/knative-plugin/src/actions/getEventPubSubActions.ts:
```diff
@@ -0,0 +1,17 @@
+
+  if (model.kind === EventingBrokerModel.kind) {
+    commonActions = Kebab.factory.common.filter((action) => {
+      return !['Delete', 'Edit'].includes(action.name);
```
@karthikjeeyar will this filter out `delete` and `edit` for broker?

karthikjeeyar replied:
Yes, this is to filter those two common actions

invincibleJai replied:
we don't need to filter karthik , we decided to have these as common actions based on user permissions i.e edit label/annotations are essentially edit broker

invincibleJai commented on frontend/packages/knative-plugin/src/components/pub-sub/PubSub.tsx:
```diff
@@ -0,0 +1,87 @@
+import { EventingBrokerModel, EventingTriggerModel, EventingSubscriptionModel } from '../../models';
+import { pubsubValidationSchema } from './pubsub-validation-utils';
+
+export interface PubSubProps {
```
nit: can remove export here

invincibleJai commented on frontend/packages/knative-plugin/src/topology/create-connector-utils.ts:
```diff
@@ -0,0 +1,49 @@
+  menu: MenuOptions,
+  connectorSource?: Node,
+): MenuOptions => {
+  if (isEventingChannelResourceKind(connectorSource?.getData().data.kind)) {
```
nit : call `isEventingChannelResourceKind` only if `connectorSource` is defined

karthikjeeyar replied:
Fixed all the comments, including filter section issue while the error message is set.

invincibleJai commented on frontend/packages/knative-plugin/src/components/pub-sub/PubSub.tsx:
```diff
@@ -0,0 +1,100 @@
+    metadata: { namespace, name: sourceName },
+  } = source;
+  const {
+    apiVersion: tApiVersion = '',
```
nit: let's call it `targetApiVersion`

invincibleJai commented on frontend/packages/knative-plugin/src/components/pub-sub/PubSubModal.tsx:
```diff
@@ -0,0 +1,54 @@
+  dirty,
+  errors,
+}) => {
+  return (
```
nit: can do a implicit return

invincibleJai commented on frontend/packages/knative-plugin/src/topology/create-connector-utils.ts:
```diff
@@ -0,0 +1,52 @@
+  if (isEventingChannelResourceKind(connectorSource?.getData().data.kind)) {
+    return [addSubscription(EventingSubscriptionModel, connectorSource.getData().resource)];
+  }
+  switch (connectorSource?.getData().data.kind) {
```
nit: we can avoid `connectorSource?` now

rachael-phillips commented:
LGTM!

invincibleJai commented:
/lgtm
/approve

invincibleJai commented:
Verified , works as expected!!

karthikjeeyar commented:
/retest
